addon/utils/query-params.ts

```diff
@@ -1,107 +1,135 @@
+ * to get filter values, along with additional passed in configuration. Optimized for JSON-API spec APIs by default.
+ * @param {buildQueryParamsArgs} Config
+ * @returns {Object} Object with query params to send to server
  */
-export function buildQueryParams(
-    controller: PaginationController,
-    offset: number = 0,
-    limit: number = 10,
-    queryParamListName: string = 'serverQueryParams',
-    includeListName: string = 'include'
-) {
-    const filterList: string[] = controller[queryParamListName];
-    const includeList: string[] = controller[includeListName];
-    let queryParams = getParamsObject(filterList, controller);
+export function buildQueryParams({
+    context,
+    offset = 0,
+    limit = 20,
+    filterList = [],
+    includeList = [],
+    sorts = [],
+    pagingRootKey = 'page',
+    filterRootKey = 'filter',
+    includeKey = 'include',
+    sortKey = 'sort',
+    serverDateFormat = 'YYYY-MM-DDTHH:mm:ss',
+    processQueryParams = (params: any) => params
+}: buildQueryParamsArgs) {
+    let queryParams = getParamsObject({
+        context,
+        filterList,
+        filterRootKey,
+        sorts,
+        sortKey,
+        serverDateFormat
+    });
+
     let pagingRoot = queryParams;
 
-    if(controller.pagingRootKey) {
-        queryParams[controller.pagingRootKey] = {};
-        pagingRoot = queryParams[controller.pagingRootKey];
+    if(pagingRootKey) {
+        queryParams[pagingRootKey] = {};
+        pagingRoot = queryParams[pagingRootKey];
     }
 
-    pagingRoot.offset = getWithDefault(controller, 'offset', offset);
-    pagingRoot.limit = getWithDefault(controller, 'limit', limit);
+    pagingRoot.offset = offset;
+    pagingRoot.limit = limit;
 
     if(isArray(includeList) && !isEmpty(includeList)) {
-        queryParams[controller.includeKey] = includeList.join(',');
+        queryParams[includeKey] = includeList.join(',');
     }
 
-    return removeEmptyQueryParams(queryParams);
+    const compactQueryParams = removeEmptyQueryParams(queryParams);
+    const processedQueryParams = processQueryParams(compactQueryParams);
+    return processedQueryParams;
 }
 
 /**
- * Gets the parameter values from the controller
- * @param parameters - The array of string names to go get from the controller
- * @param context - The pagination controller instance
- * @returns - Object with query params to send to server
+ * Gets the parameter values from the "context" (e.g. Controller/Component)
+ * @param {getParamsObjectArgs} Config
+ * @returns {Object} Object with query params to send to server
  */
-export function getParamsObject(parameters: string[] | undefined, context: PaginationController) {
+export function getParamsObject({
+    context,
+    filterList,
+    filterRootKey,
+    sorts = [],
+    sortKey = 'sort',
+    serverDateFormat = 'YYYY-MM-DDTHH:mm:ss'
+}: getParamsObjectArgs) {
     let params: any = {};
```

Comment:
I feel like we could probably type this a bit better - at the very least so its an object that takes any string key with a value